src/main/scala/com/fulcrumgenomics/vcf/DownsampleVcf.scala

```diff
@@ -0,0 +1,281 @@
+
+      def isInOrder(current: Variant, next: Variant, currentIndex: Int, nextIndex: Int): Boolean = {
+        (currentIndex < nextIndex) || (currentIndex == nextIndex && current.end <= next.pos)
+      }
+
+      def next(): Variant = {
+        val current = iter.next()
+        val currentIndex = dict(current.chrom).index
+        iter.dropWhile { next: Variant =>
+          val nextIndex = dict(next.chrom).index
+          require(
+            isInOrder(current, next, currentIndex, nextIndex),
+            f"variants out of order; ${current.chrom}:${current.pos} > ${next.chrom}:${next.pos}")
+
+          currentIndex == nextIndex && next.pos - current.end < windowSize
+        }
+        current
+      }
+    }
+  }
+
+  /** Downsamples variants using Allele Depths
+   *
+   * @param oldAds     an indexed seq of the original allele depths
+   * @param proportion the proportion to use for downsampling,
+   *                   calculated using total base count from the index and a target base count
+   * @return a new IndexedSeq of allele depths of the same length as `oldAds`
+   */
+  def downsampleADs(oldAds: IndexedSeq[Int], proportion: Double, random: Random): IndexedSeq[Int] = {
+    require(proportion <= 1, f"proportion must be less than 1: proportion = ${proportion}")
+    oldAds.map(s => Range(0, s).iterator.map(_ => random.nextDouble()).count(_ < proportion))
+  }
+
+  /**
+   * Does the downsampling on a Variant
+   * @param variant the variant with the genotype to downsample
+   * @param proportions a map of downsampling target proportions for each sample
+   * @param random random number generator for downsampling
+   * @param epsilon the error rate for genotyping
+   * @return a new variant with updated genotypes
+   */
+  // Returns a new variant that has downsampled ADs, recomputed PLs and updated genotypes
+  def downsampleAndRegenotype(variant: Variant, proportions: Map[String, Double], random: Random, epsilon: Double = 0.01): Variant = {
+    try {
+      variant.copy(genotypes = variant.genotypes.map { case (sample, gt) =>
+        val proportion = proportions(sample)
+        sample -> downsampleAndRegenotype(gt = gt, proportion = proportion, random = random, epsilon = epsilon)
+      })
+    } catch {
+      case e: MatchError => throw new Exception(
+        "processing " + variant.id +  " at " + variant.chrom + ":" + variant.pos + "-" + variant.end, e
+      )
+    }
+  }
+
+  /**
+   * Does the downsampling on a Genotype
+   * @param gt the genotype to downsample
+   * @param proportion the proportion to use for downsampling allele depths
+   * @param random random number generator for downsampling
+   * @param epsilon the error rate for genotyping
+   * @return a new Genotype with updated allele depths, PLs and genotype
+   */
+  def downsampleAndRegenotype(gt: Genotype, proportion: Double, random: Random, epsilon: Double): Genotype = {
+    val oldAds = gt[IndexedSeq[Int]]("AD")
```

Comment:
consider a `getOrElse[IndexedSeq[Int]]("AD", throw new Exception("AD not found for sample..."))` (this is pseudocode)